From e2b9f51f7f88aa406e7a653204eea5bb4fce5880 Mon Sep 17 00:00:00 2001
From: Tomas Henzl <thenzl@redhat.com>
Date: Mon, 8 Sep 2014 14:32:11 -0400
Subject: [scsi] hpsa: remove online devices from offline device list

Message-id: <1410186734-19986-81-git-send-email-thenzl@redhat.com>
Patchwork-id: 92103
O-Subject: [RHEL6.6 e-stor PATCH 80/83] hpsa: remove online devices from offline device list
Bugzilla: 1069185
RH-Acked-by: Maurizio Lombardi <mlombard@redhat.com>
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Ewan Milne <emilne@redhat.com>

When devices come on line, they should be removed from the list of
offline devices that are monitored.

Signed-off-by: Stephen M. Cameron <scameron@beardog.cce.hp.com>
Reviewed-by: Scott Teel <scott.teel@hp.com>
Reviewed-by: Joe Handzik <joseph.t.handzik@hp.com>
Reviewed by: Mike MIller <michael.miller@canonical.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>

(cherry picked from commit d1fea47c36c13df5fc7e5a379ed5c4d2059de583)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/hpsa.c b/drivers/scsi/hpsa.c
index 31d2614..b131813 100644
--- a/drivers/scsi/hpsa.c
+++ b/drivers/scsi/hpsa.c
@@ -6940,8 +6940,12 @@ static int hpsa_offline_devices_ready(struct ctlr_info *h)
   d = list_entry(this, struct offline_device_entry,
     offline_list);
   spin_unlock_irqrestore(&h->offline_device_lock, flags);
-  if (!hpsa_volume_offline(h, d->scsi3addr))
+  if (!hpsa_volume_offline(h, d->scsi3addr)) {
+   spin_lock_irqsave(&h->offline_device_lock, flags);
+   list_del(&d->offline_list);
+   spin_unlock_irqrestore(&h->offline_device_lock, flags);
    return 1;
+  }
   spin_lock_irqsave(&h->offline_device_lock, flags);
  }
  spin_unlock_irqrestore(&h->offline_device_lock, flags);
-- 
1.7.1